File: 0744. Find Smallest Letter Greater Than Target/0744.py
# time complexity: O(logn) - each recursive call halves the search range
# space complexity: O(logn) - recursion stack depth proportional to logn
class Solution(object):
    # recursive binary search
    def nextGreatestLetter(self, letters, target):
        """
        :type letters: List[str]
        :type target: str
        :rtype: str
        """
        # recursive helper returns the index of the first letter > target
        def helper(low, high):
            # base case: search interval is over
            if low > high:
                return low
            mid = low + (high-low)//2
            if letters[mid] <= target:
                # search right half if mid is not greater
                return helper(mid+1, high)
            else:
                # search left half if mid is greater
                return helper(low, mid-1) 

        idx = helper(0, len(letters)-1)

        if idx == len(letters):
            # if no greater letter found
            return letters[0]
        
        return letters[idx]
        
        # alternative way for 3 lines above
        # return letters[idx % len(letters)]

File: 0744. Find Smallest Letter Greater Than Target/0744.java
// time complexity: O(n)
// space complexity: O(1)
class Solution {
    // linear scan
    public char nextGreatestLetter(char[] letters, char target) {
        // letters is sorted in non-decreasing order
        for(int i=0; i<letters.length; i++){
            // return first character greater than target
            if(letters[i] > target){
                return letters[i];
            }
        }
        // if none found, return the first element
        return letters[0];
    }
}